### ISSUE TITLE
Expected number of samples from load()?

### ISSUE THREAD

**Issue by sammlapp** [librosa/librosa#1709]
When duration*sample_rate is not an integer, it seems that the behavior of __soundfile_load and __audioread_load are different with respect to how many samples will be returned: 

__soundfile_load performs int(duration*sample), ie a 'floor' operation:

https://github.com/librosa/librosa/blob/f21bda6b9fd7d9f1276749862247a6daf59e792d/librosa/core/audio.py#LL215C13-L215C13

while __audioread_load performs int(np.round(duration*sample)): 

https://github.com/librosa/librosa/blob/f21bda6b9fd7d9f1276749862247a6daf59e792d/librosa/core/audio.py#LL247C9-L247C9

Is this behavior intentional? 

Furthermore, neither of these behaviors seem to explain observed results: (sample rate is 22050) 
```
samples, sr = librosa.load(path=audio_path,offset=0,duration=5/22050) #5 samples
samples, sr = librosa.load(path=audio_path,offset=0,duration=5.1/22050) #5 samples
samples, sr = librosa.load(path=audio_path,offset=0,duration=5.9/22050) #6 samples
samples, sr = librosa.load(path=audio_path,offset=0,duration=5.51/22050) #5 samples (?!)
```
Is this a floating point precision issue? `int(np.round(5.51/22050*22050))` gives me 6, not 5, so I don't know why this last line gives 5 samples.

**Comment by bmcfee** [comment]
I think there are a few things going on here.

TLDR: yes, this does appear to be somewhat inconsistent, and not intentional.  However, the lines you're linking to refer to the `offset` calculation, which will be bypassed in your example.

In your example code, it's not obvious or controlled whether you're hitting the audioread or soundfile code path, so I can't say for certain why you're getting different results in the last two lines.  You also state that the sample rate is 22050, but is that guaranteed?  If the native rate of the file is something else, then you may hit rounding errors in resampling that could be producing the different length behaviors.

At this point, my preference would be for the floor style, especially since the long-term plan is to deprecate audioread anyway #1456 .  This would mean that the correct lengths for your examples should be 5 samples in all cases, but only if we assume that the native sample rate is 22050.

**Comment by bmcfee** [comment]
Flagging this for 0.10.1 to implement flooring consistently across decoders.

**Comment by bmcfee** [comment]
Looking into this one today.  We also have a bit of disagreement in how `duration` and `offset` are handled by our signal generators.  Tone and chirp take `int(ceil(...))`; soundfile load and stream take floor, audioread and clicks take round.

I propose that we adjust all of these to use floor instead.  This way, the `duration` parameter would be consistently* interpreted across all functions.

*: not exactly.  The load functions apply truncation before sample rate conversion, so there's still a chance for a bit of rounding discrepancies depending on which resampler is used.  I don't think there's anything we can do about that, short of truncating after resampling.

**Comment by bmcfee** [comment]
.. and for the sake of consistency, `time_to_samples` uses integer casting, which is implicitly flooring `time * sr`.

**Comment by sammlapp** [comment]
Thanks for looking into this and addressing it!
